common/formats/src/format.rs

```diff
@@ -45,6 +45,10 @@ pub trait InputFormat: Send + Sync {
 
     fn read_buf(&self, buf: &[u8], state: &mut Box<dyn InputState>) -> Result<usize>;
 
+    fn set_buf(&self, buf: Vec<u8>, state: &mut Box<dyn InputState>);
```

Comment:
> BTW, will this API replace `read_buf`?

 now it is used in other places.

Reply:
> How about adding a detailed behavior for this API?
> 
> * Which behavior is expected if `state.memory` has data?
> * Is there any requirement for input values?
>   
>   * Can `buf.len()` be `0`?
>   * Input `state` must be initiated correctly, or will we panic?

this code is just to reuse some code with InputState/InputFormat (mainly read_buf)

InputState.memory is originally designed to be filled by read_buf.

set_buf is called on a new/empty InputState.

I want to refactor InputState later. since  it is used in other places with diff considerations.

Reply:
> I want to refactor InputState later. since it is used in other places with diff considerations.

Nice idea!

Please create a tracking issue for this work and link here.